### ISSUE TITLE
[DAG] Failure to reassociate SMAX/SMIN/UMAX/UMIN

### ISSUE THREAD

**Issue by RKSimon** [llvm/llvm-project#58110]
While opt will reassociate IMINMAX intrinsics, if they appear later in DAG we fail to handle them:
```
declare i32 @llvm.umin.i32(i32, i32)
define i32 @foo(i32 %x) {
  %a = call i32 @llvm.umin.i32(i32 %x, i32 17)
  %b = call i32 @llvm.umin.i32(i32 %a, i32 17) 
  ret i32 %b
}
```
llc -mtriple=x86_64--
```
foo:                                    # @foo
        cmpl    $17, %edi
        movl    $17, %eax
        cmovael %eax, %edi
        cmpl    $17, %edi
        cmovbl  %edi, %eax
        retq
```
Godbolt: https://gcc.godbolt.org/z/45sYa7xj1

This is mainly about adding test coverage - the actual code change to add a reassociateOps call to visitIMINMAX is trivial

**Comment by vfdff** [comment]
we may already address this in some other pass, https://gcc.godbolt.org/z/nreY96nsx

**Comment by mbelicki** [comment]
I'll take a look and provide a patch today/tomorrow.

**Comment by RKSimon** [comment]
> we may already address this in some other pass, https://gcc.godbolt.org/z/nreY96nsx

Yes, however we often generate additional IMIN/MAX nodes inside the DAG as well after the middle passes

**Comment by RKSimon** [comment]
Candidate Patch: https://reviews.llvm.org/D142284

**Comment by wanghuibin0** [comment]
Has this bug been fixed?

**Comment by RKSimon** [comment]
No, the previous patch (while fixing the issue) was missing some extra test coverage that proved tricky to reduce properly. Plus I felt that we should probably see if we can reduce the increase in instructions in fixed-vector-trunc-vp.ll first somehow.

But feel free to fork the old phab patch, check the current state and create a PR

**Comment by SahilPatidar** [comment]
@RKSimon , 
I would like to work on this one, Please let me know if there are any specific guidelines or procedures I should follow.

**Comment by SahilPatidar** [comment]
Hello @RKSimon, @dtcxzyw,

I've reviewed your commit and tested the changes on the test case you mentioned. It appears to be working fine as expected. However, I'm struggling to understand the exact issue we're trying to address. Could you please provide a brief description of the current problem we're facing?

Thank you.

**Comment by RKSimon** [comment]
Is the fixed-vector-trunc-vp.ll regression still there?

I can't find the clamping source code test which was the original reason that I started this 😞

**Comment by SahilPatidar** [comment]
@RKSimon,

I recently updated the fixed-vectors-trunc-vp.ll file with the code you committed. After the update, I noticed some changes in the registers. Could you please explain the reason behind these modifications?

Thank you!

**Comment by RKSimon** [comment]
Maybe create a draft PR?

**Comment by SahilPatidar** [comment]
@RKSimon,
here is Draft PR.
